### ISSUE TITLE
Implement more functions for AirPods by communcating to them like an apple device

### ISSUE THREAD

**Issue by kavishdevar** [TheParasiteProject/packages_modules_Bluetooth_android_BtHelper#1]
https://github.com/Project-PixelStar/packages_apps_BtHelper/issues/1

> Hey!
> 
> I've managed to reverse engineer airpods functionalities and how they work on apple devices. Currently there's a bug in the AOSP bluetooth stack that doesn't allow it to work though.
> 
> Apple uses the L2CAP protocol to communicate with the airpods, they use the PSM 4097 or 0x1001. But the problem with using this is that in flouride, it sends the airpods a packet to check for extended flow control, and wait for the device to respond. But the airpods don't respond to anything until a specific handshake packet is sent first. So android just doesn't continue with the connection.. I commented out the lines in the bluetooth stack and replaced the `libbluetooth_jni.so` (Ask me more about it), and voilà! It worked.. Instead of relying on the advertisements, the btservices can now communicate with the airpods just like an apple device would!! ear-detection, accurate battery percentages, setting anc modes, controlling what the button does, accessiblity settings, controlling adaptive transparency strength (new on app2) can be implemented easily!
> 
> But the only problem is that I can build the BtServices app, but it being a soong-built app, takes a long time to build.. and isn't ideal for experimenting. But, I might end up making some changes and making a PR, no promises though.

**Comment by someone5678** [comment]
Thank you so much for issue! I'll check it later!

**Comment by someone5678** [comment]
Okay, so, I also tried to implement these features before but failed. So if you don't mind, could you please explain me more about

> commented out the lines in the bluetooth stack and replaced the libbluetooth_jni.so (Ask me more about it),

this section? Thanks!!

**Comment by kavishdevar** [comment]
i've commented this [line](https://cs.android.com/android/platform/superproject/+/android14-qpr3-release:packages/modules/Bluetooth/system/stack/l2cap/l2c_link.cc;l=118)

and [this](https://cs.android.com/android/platform/superproject/+/android14-qpr3-release:packages/modules/Bluetooth/system/stack/l2cap/l2c_fcr.cc;l=1605-1610) check performed for extended flow too.

This fixes the issue where android can't open l2c connections with airpods. Then i built the `libbluetooth_jni.so` using `make libbluetooth_jni` and replaced it in the lib directories using a script in post-fs-data (KSU) that creates an overlayfs for `/apex/com.android.btservices` and `/apex/com.android.btservices@350090000`. But this renders bluetooth unsable probably because of some signature verification, so i made a module that adds the library in `/system/lib64`. (finally worked!).

you can also check this commit - https://github.com/kavishdevar/packages_modules_Bluetooth/commit/81afc3f

also, i'm a noob at platform developement, but i can try writing java code, only if there is a quicker way to build apks without using `mmm .` or `make BtHelper`?

**Comment by kavishdevar** [comment]
this of course is a temporary fix, maybe the send function shouldn't block if it is sending this particular packet, and should return false if no response is received from the peer. i really want to create an issue on android, but i am too lazy/busy with school!

If you'd like to implement the actual airpods functions, checkout my very quick and dirty code for linux at https://github.com/kavishdevar/aln (check my main start-daemon script there, it has battery and ear detection parsing, and the enums for other stuff like setting anc), and https://github.com/steam3d/MagicPodsCore/ either the `decky` branch (originally written in c, probably has more stuff implemented) or the `main` branch (a rewrite in python)

**Comment by someone5678** [comment]
Thanks! I'm currently in build and testing stage!

**Comment by someone5678** [comment]
> make BtHelper

Yes.. On AOSP, `make <module-name>` is only way to build that module only...
You can pass -j argument though like `m -j$(nproc) BtHelper`

**Comment by kavishdevar** [comment]
> Thanks! I'm currently in build and testing stage!

Please share the apk once built, if you can :)

**Comment by kavishdevar** [comment]
> > make BtHelper
> 
> Yes.. On AOSP, `make <module-name>` is only way to build that module only...
> You can pass -j argument though like `m -j$(nproc) BtHelper`

I already use the job argument, but, my pc is under-speced, takes at least 20 minutes to build.. :(

I'll try on my dad's pc later.

**Comment by kavishdevar** [comment]
I've opened an issue on issuetracker, please upvote :)

https://issuetracker.google.com/issues/371713238

**Comment by kavishdevar** [comment]
hey! what's the best ide to develop this app? and are you planning to implement everything, if so, then please lemme know, i can help with the l2c packets and all!

**Comment by someone5678** [comment]
> hey! what's the best ide to develop this app? and are you planning to implement everything, if so, then please lemme know, i can help with the l2c packets and all!

Probably Android Studio!
But as this is system app, requires some modification just like LineageOS' Updater app does.
e.g.) [LineageOS updater](https://github.com/search?q=repo%3ALineageOS%2Fandroid_packages_apps_Updater+android+studio&type=commits)

And yes. I'm planning to implement like what I saw in those Decky loader plugin does (ANC, etc.)

If you don't mind, I'll really appreciate your help!!

Currently, I'm testing your changes and some modification!
Feel free to pr or issue anytime!

**Comment by kavishdevar** [comment]
well, i am not planning to pr just because my pc is too slow to build and test, but if i can use gradle, then, absolutely in for this!

> If you don't mind, I'll really appreciate your help!!

I absolutely do not :D, if you need, I can list out all the AAP (Apple Accessory Protocol)  packets sent/received.. (you can also check my aln repo instead, but it might be a pain to navigate through)

btw, we can use IM platform if you'd prefer that over gh issues for quick chats

**Comment by someone5678** [comment]
> well, i am not planning to pr just because my pc is too slow to build and test, but if i can use gradle, then, absolutely in for this!
> 
> > If you don't mind, I'll really appreciate your help!!
> 
> I absolutely do not :D, if you need, I can list out all the AAP (Apple Accessory Protocol) packets sent/received.. (you can also check my aln repo instead, but it might be a pain to navigate through)
> 
> btw, we can use IM platform if you'd prefer that over gh issues for quick chats

Okay thanks!!

As a quick check, your aln repo definitely helps my dev for BtHelper. XD

For IM, I only have telegram account eh...

**Comment by kavishdevar** [comment]
Also, here's a list of features i've implemented (should add this to my repo too 😅)

- [x] Receive Battery Information
- [x] Set/Receive ANC Modes
- [x] Set Adaptive Audio Noise settings
- [x] Receive In-Ear detection Status 
- [ ] Personalized Volume (idk how this works, if it received data from icloud, or is purely from airpods)
- [x] Conversational Awareness
- [x] Ear Detection
- [ ] Head Gestures
- [ ] Siri (Voice assistant on long stem press)
- [ ] Hold and Press configuration (this is really weird, mac sends different packets based on what the current status is, instead of a fixed packet for what the current button state is, like common there's only so many to map out numbers to states)
- [ ] Head Tracking (i really want this, could be easy, could be difficult, i'll never know because i don't have a device with apple silicon 😭)
- [x] Case Charging Sounds
- [x] Rename AirPods
- [ ] Mute Unmute Calls
- Accessibilty
    - [ ] Press Speed
    - [ ] Press and hold duration
    - [ ] Noise Cancellation with one AirPod
    - [ ] Tone Volume
    - [ ] Toggle Volume Control on Swipe (APP only, i believe)
    - [ ] Volume Swipe (Normal/Longer/Longest)
    - [ ] Headphone accomodation (I literally can't tell the difference between any samples played, lol, also, idk if this is something that the mac does)
    - [ ] Audio Tuning (idk if this is also smth that mac does)
    - [ ] Customize Transparency Mode (This is gonna take some while to parse, it is 103 bytes :(... probably all the 4 sliders and 1 switch under this is sent as a whole)

**Comment by kavishdevar** [comment]
> As a quick check, your aln repo definitely helps my dev for BtHelper. XD

Glad my repo is useful for someone other than me :]

> For IM, I only have telegram account eh...

i do have telegram too, _username removed for privacy_
